.github/workflows/ci.yml

```
@@ -1,70 +1,144 @@
 name: CI
 
-# Trigger the workflow on push or pull request, but only for the main branch
 on:
   pull_request:
+    types: [synchronize, opened, reopened]
   push:
     branches: [main]
+  schedule:
+    # additionally run once per week (At 00:00 on Sunday) to maintain cache
+    - cron: '0 0 * * 0'
+
 
 jobs:
-  build:
-    name: ghc ${{ matrix.ghc }}
-    runs-on: ubuntu-16.04
+  cabal:
+    name: ${{ matrix.os }} / ghc ${{ matrix.ghc }}
+    runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        cabal: ["3.2"]
+        os: [ubuntu-latest, macOS-latest, windows-latest]
+        cabal: ["3.4"]
         ghc:
           - "8.4.4"
           - "8.6.5"
           - "8.8.4"
-          - "8.10.3"
+          - "8.10.4"
+          - "9.0.1"
+        exclude:
+          - os: macOS-latest
+            ghc: 8.10.4
+          - os: macOS-latest
+            ghc: 8.8.4
+          - os: macOS-latest
+            ghc: 8.6.5
+          - os: macOS-latest
+            ghc: 8.4.4
+
+          - os: windows-latest
+            ghc: 8.10.4
+          - os: windows-latest
+            ghc: 8.8.4
+          - os: windows-latest
+            ghc: 8.6.5
+          - os: windows-latest
+            ghc: 8.4.4
 
     steps:
-    - uses: actions/checkout@v2.3.3
-      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/main'
+    - uses: actions/checkout@v2
 
     - uses: haskell/actions/setup@v1
+      id: setup-haskell-cabal
       name: Setup Haskell
       with:
         ghc-version: ${{ matrix.ghc }}
         cabal-version: ${{ matrix.cabal }}
 
-    - uses: actions/cache@v2.1.2
+    - name: Configure
+      run: |
+        cabal configure --enable-tests --enable-benchmarks --enable-documentation --test-show-details=direct --write-ghc-environment-files=always
+
+    - name: Freeze
+      run: |
+        cabal freeze
+
+    - uses: actions/cache@v2.1.4
       name: Cache ~/.cabal/store
       with:
-        path: ~/.cabal/store
-        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal
+        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
+        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
+
+    - name: Install dependencies
+      run: |
+        cabal build all --only-dependencies
 
     - name: Build
       run: |
-        cabal v2-update
-        cabal v2-build all --enable-tests --enable-benchmarks
+        cabal build all
 
     - name: Unit and property tests
       run: |
-        cabal v2-test all --enable-tests --test-show-details=direct
+        cabal test all --enable-tests --test-show-details=direct
```

I think this is not needed as it's already in `configure`
```suggestion
        cabal test all
```